pkg/storage/etcd3/watcher.go

```diff
@@ -182,6 +187,11 @@ func (wc *watchChan) startWatching() {
 			wc.sendEvent(parseEvent(e))
 		}
 	}
+	// When we come to this point, it's only possible that client side ends the watch.
+	// e.g. cancel the context, close the client.
+	// If this watch chan is broken and context isn't cancelled, other goroutines will still hang.
+	// We should notify the main thread that this goroutine has exited.
+	close(watchClosedCh)
```

Comment:
ok, I still think this needs a lot of changes to be understandable, but I won't block this any more.